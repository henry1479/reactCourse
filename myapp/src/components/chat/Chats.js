import { List,ListItem,ListItemText} from '@material-ui/core';
import { NavLink } from 'react-router-dom';


export const chatsBlockTests = {
    chatName: 'chatBlock-name',
    loginField: 'chatBlock-loginField'
}

export const Chats = ( props) => {
    return (
        <div style={{
            float:'left',
            width:'40%',
            border: '1px solid #ccc',
            background: '#dda0dd',
            marginBottom: '5%',
            marginLeft: '5%',
            borderRadius: '15px'
        }}>
            <List>
                <h3>Chat List</h3>
                {props.chats?.map((el) => (
                    <ListItem key={el.id} >
                        <NavLink to={`/chats/${el.id}`} style={{textDecoration: 'none',}} activeStyle={{color: 'red', textDecoration: 'none',}}>
                            <ListItemText data-testid ={chatsBlockTests.chatName} >
                                {el.chat}
                            </ListItemText>
                        </NavLink>
                    </ListItem>
                    )
                )   
                }
            </List>
            <form action="#" className="chatlist__form">
                <label htmlFor="name">Enter the name of new friend</label>
                <input type="text" name="friend-name" className="chatlist__input" onChange={props.handleChange} value={props.chat} data-testid={chatsBlockTests.loginField}/>
                <input type="button" value="Add" className="chatlist__button" onClick={props.onAddChatFirebase}/>
                <input type="button" value="Remove" className="chatlist__button" onClick={props.onRemoveChat}/>
            </form>
        </div>
    )
}