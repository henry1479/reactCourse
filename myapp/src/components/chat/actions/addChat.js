export const ADD_CHAT = 'CHAT::ADD_CHAT';

export const addName = (newName) => ({
    type: ADD_CHAT,
    payload: newName,
})
