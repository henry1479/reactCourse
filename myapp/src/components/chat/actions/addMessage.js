export const ADD_MESSAGE = 'CHAT::ADD_MESSAGE';

export const setMessage = (newMessage) => ({
    type: ADD_MESSAGE,
    payload: newMessage
})

