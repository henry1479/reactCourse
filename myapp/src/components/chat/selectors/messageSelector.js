export const getMessageList = (chatId) => state => state.chatListReducer.chats[chatId]