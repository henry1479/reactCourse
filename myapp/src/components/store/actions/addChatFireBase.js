import { db } from '../../../index.js'


export const CHANGE_CHATS ='CHANGE_CHATS';


const getPayloadFromSnapshot = (snapshot) => {
    const chats = [];
    snapshot.forEach(entry => {
    chats.push({id: entry.key, ...entry.val()})
    });

   
   
    return chats;
  }
  
  export const addChatWithFirebase = (chat) => async () => {
    db().ref('chats').push(chat);
   
    
  };
  
  export const initChatTracking = () => (dispatch) => {
    // db().ref("chats").on("child_changed", (snapshot) => {
    //   const payload = getPayloadFromSnapshot(snapshot);
    //   dispatch({
    //     type: CHANGE_CHATS,
    //     payload,
    //   });
    // });
  
    db().ref("chats").on("value", (snapshot) => {
      const payload = getPayloadFromSnapshot(snapshot);
      dispatch({
        type: CHANGE_CHATS,
        payload,
      });
    });
  
    // db().ref("chats").on("child_added", (snapshot) => {
    //   const payload = getPayloadFromSnapshot(snapshot);
    //   dispatch({
    //     type: CHANGE_CHATS,
    //     payload,
    //   });
    // });
  };