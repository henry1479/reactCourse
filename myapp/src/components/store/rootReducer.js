import { combineReducers } from 'redux';
import chatListReducer from './reducers/chatListReducer';
import  profileReducer  from './reducers/profileReducer';
import postsPageReducer from './reducers/postsPageReducer';
import {messagesReducer} from './reducers/firebaseReducer';

// объеденяем редьюсеры
export const rootReducer = combineReducers({
    profileReducer,
    chatListReducer,
    postsPageReducer,
    messagesReducer
})




