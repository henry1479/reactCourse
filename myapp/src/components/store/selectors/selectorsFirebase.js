export const getMessageListFirebase = state => state.messagesReducer.messages

export const getChatFirebase = state => state.messagesReducer?.chats