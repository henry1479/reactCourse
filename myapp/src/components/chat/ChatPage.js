
import React from 'react';
import {  useDispatch,useSelector } from 'react-redux';
import { db } from '../../index.js';
import { useState, useCallback, useEffect } from 'react';
import { addChatWithFirebase } from '../store/actions/addChatFireBase';
import { initChatTracking } from '../store/actions/addChatFireBase';
import { Chats }  from './Chats.js';
import MessageList from './MessageList.js';
import { Route,useParams} from 'react-router-dom';
import { getChatFirebase } from '../store/selectors/selectorsFirebase';

const ChatsPure = React.memo(Chats);

//реализует список чатов с формой для новых и списком сообщений
// имеет собственную функцию по отправке новых чатов в fb
// и получает их из Redux с помощью initChatTracking()

function ChatPage() {
    // селектор для получения чатов из редьюсера
    const chats  = useSelector(getChatFirebase);
    // стэйт для получения данных из поля ввода
    const [ chat, setChat] = useState('');
    let { Id } = useParams();
    const dispatch = useDispatch(); 
    

    // получает данные из поля ввода и оправляет в стэйт
    const handleChange = useCallback((event)=>{  
        let value = event.target.value;
        setChat(value);
    },[Id]);

    // отправляет чат в fb
    const onAddChatFirebase = useCallback(
        () => {
            dispatch(
                addChatWithFirebase({chat})
            )
           
          },
          [chat,dispatch]
        );
 
    
    // направляет данные по чатам чаты из fb в редьюсер через мидлвэр
    useEffect(() =>{
        dispatch(initChatTracking())},
        [dispatch]
    )
    
    // удаляет выбранный чат прямо из fb
    const onRemoveChat = useCallback(()=> {
        db().ref('chats').child(Id).remove()
    },[Id]);


    // если все чаты удалены, то путь к чатам - пустая строка
    if(chats.length === 0) {
        Id = " ";
    }

        
    return (
        <>
            <ChatsPure 
            onAddChatFirebase={onAddChatFirebase}
            onRemoveChat={onRemoveChat}
            id={Id}
            chat={chat}
            handleChange={handleChange} 
            chats={chats}
            />
            <Route path={`/chats/${Id}`}>
                <MessageList id={Id} />
            </Route>
        </>

    );
}
    
  

export default ChatPage;