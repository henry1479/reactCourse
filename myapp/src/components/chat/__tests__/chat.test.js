import { Provider } from 'react-redux';
import { render, fireEvent, waitFor, screen, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Chats, chatsBlockTests } from '../Chats';
import { BrowserRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ChatPage from '../ChatPage';
import { CHANGE_CHATS } from '../../store/actions/addChatFireBase';
import { getChatFirebase } from '../../store/selectors/selectorsFirebase';
import { createStore } from 'redux';
import { rootReducer } from '../../store/rootReducer';




 





describe('Chats', () => {
    it('вывод текста ссылки в списке чатов из свойства chat объекта в массиве, переданного в комопонент из ChatPage через пропс chats ', ()=> {
        let myMock = jest.fn();

        myMock.mockReturnValue([{chat:'Kostya', id:'122323'}]);
        const chats = myMock();

        const component= render(<BrowserRouter><Chats chats={chats}/></BrowserRouter>);

        const received = component.getByTestId(chatsBlockTests.chatName).firstChild
        const [ expected ] = component.getAllByText(chats[0].chat);

        expect(received).toEqual(expected)

    }); 

    it('вызов обработчика handleChange,переданного через пропс из ChatPage,  на изменение значения поля ввода для имени чата', () => {
        const chat = "Vasya";
        const handleChange = jest.fn();
        const component = render(<BrowserRouter><Chats handleChange={handleChange} /></BrowserRouter>);
        fireEvent.change(component.getByTestId(chatsBlockTests.loginField), {target: {value: chat}});
        expect(handleChange).toBeCalled();
    })

})


describe('ChatPage', () => {
    it ('компонент диспачит экшен в стор для получения списка чатов из Firebase',  () => {
        const store = createStore(rootReducer);
        render(<Provider store={store}><ChatPage /></Provider>)
        const actions = store.getActions();
        const lastAction = actions[actions.length - 1];
        expect(lastAction).toEqual(getChatFirebase());
    });
    
})





