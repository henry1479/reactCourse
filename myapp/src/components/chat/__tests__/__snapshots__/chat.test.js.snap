// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chats сравнение снэпшотов 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="float: left; width: 40%; border: 1px solid #ccc; background: rgb(221, 160, 221); margin-bottom: 5%; margin-left: 5%; border-radius: 15px;"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <h3>
            Chat List
          </h3>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <a
              href="/chats/1234"
              style="text-decoration: none;"
            >
              <div
                class="MuiListItemText-root"
                data-testid="chatBlock-name"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  Petya
                </span>
              </div>
            </a>
          </li>
        </ul>
        <form
          action="#"
          class="chatlist__form"
        >
          <label
            for="name"
          >
            Enter the name of new friend
          </label>
          <input
            class="chatlist__input"
            name="friend-name"
            type="text"
            value=""
          />
          <input
            class="chatlist__button"
            type="button"
            value="Add"
          />
          <input
            class="chatlist__button"
            type="button"
            value="Remove"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="float: left; width: 40%; border: 1px solid #ccc; background: rgb(221, 160, 221); margin-bottom: 5%; margin-left: 5%; border-radius: 15px;"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <h3>
          Chat List
        </h3>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <a
            href="/chats/1234"
            style="text-decoration: none;"
          >
            <div
              class="MuiListItemText-root"
              data-testid="chatBlock-name"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
              >
                Petya
              </span>
            </div>
          </a>
        </li>
      </ul>
      <form
        action="#"
        class="chatlist__form"
      >
        <label
          for="name"
        >
          Enter the name of new friend
        </label>
        <input
          class="chatlist__input"
          name="friend-name"
          type="text"
          value=""
        />
        <input
          class="chatlist__button"
          type="button"
          value="Add"
        />
        <input
          class="chatlist__button"
          type="button"
          value="Remove"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Chats сравнение снэпшотов компонента 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="float: left; width: 40%; border: 1px solid #ccc; background: rgb(221, 160, 221); margin-bottom: 5%; margin-left: 5%; border-radius: 15px;"
      >
        <ul
          class="MuiList-root MuiList-padding"
        >
          <h3>
            Chat List
          </h3>
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <a
              href="/chats/1234"
              style="text-decoration: none;"
            >
              <div
                class="MuiListItemText-root"
                data-testid="chatBlock-name"
              />
            </a>
          </li>
        </ul>
        <form
          action="#"
          class="chatlist__form"
        >
          <label
            for="name"
          >
            Enter the name of new friend
          </label>
          <input
            class="chatlist__input"
            name="friend-name"
            type="text"
            value=""
          />
          <input
            class="chatlist__button"
            type="button"
            value="Add"
          />
          <input
            class="chatlist__button"
            type="button"
            value="Remove"
          />
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="float: left; width: 40%; border: 1px solid #ccc; background: rgb(221, 160, 221); margin-bottom: 5%; margin-left: 5%; border-radius: 15px;"
    >
      <ul
        class="MuiList-root MuiList-padding"
      >
        <h3>
          Chat List
        </h3>
        <li
          class="MuiListItem-root MuiListItem-gutters"
        >
          <a
            href="/chats/1234"
            style="text-decoration: none;"
          >
            <div
              class="MuiListItemText-root"
              data-testid="chatBlock-name"
            />
          </a>
        </li>
      </ul>
      <form
        action="#"
        class="chatlist__form"
      >
        <label
          for="name"
        >
          Enter the name of new friend
        </label>
        <input
          class="chatlist__input"
          name="friend-name"
          type="text"
          value=""
        />
        <input
          class="chatlist__button"
          type="button"
          value="Add"
        />
        <input
          class="chatlist__button"
          type="button"
          value="Remove"
        />
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
